SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -29,54 +29,61 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 template <typename DataT>
 using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
 
-// Calls simd constructor in provided invocation context, which depends on the
-// TestCaseT type. TestCaseT is a struct, that should have call_simd_ctor method
-// that should return constructed object of simd class.
-// This function returns std::vector instance with the output data.
-template <typename DataT, int NumElems, typename TestCaseT>
-auto call_simd(sycl::queue &queue, const shared_vector<DataT> &ref_data) {
-
-  shared_vector<DataT> result{NumElems, shared_allocator<DataT>{queue}};
-
-  queue.submit([&](sycl::handler &cgh) {
-    const auto ref = ref_data.data();
-    auto out = result.data();
+template <typename DataT, int NumElems = 0>
+class TestDescription : public ITestDescription {
+protected:
+  const std::string m_data_type;
+  const std::string m_test_case_description;
+  DataT m_retrieved_val;
+  DataT m_expected_val;
+  size_t m_index;
+
+public:
+  TestDescription(const std::string &test_case_description,
+                  const std::string &data_type, DataT retrieved_val,
+                  DataT expected_val, size_t index)
+      : m_data_type(data_type), m_test_case_description(test_case_description),
+        m_retrieved_val(retrieved_val), m_expected_val(expected_val),
+        m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: convert fp types into hex data type
+    std::string log_msg(
+        "Simd by " + m_test_case_description +
+        " failed, retrieved: " + std::to_string(m_retrieved_val) +
+        ", expected: " + std::to_string(m_expected_val) + ", at index: " +
+        std::to_string(m_index) + ", with data type: " + m_data_type);
+
+    if constexpr (NumElems != 0) {
+      log_msg += ", with num elements: ";
+      log_msg += std::to_string(NumElems);
+    }
```

Comment:
(1) Looks like we don't need a separate case for zero NumElems, what do you think?
(2) Minor, feel free to ignore, but by using `operator+()` we might split string construction over multiple lines and have lesser git diffs in case of changes
(3) Also minor, but we might just print `simd<T,N>` with actual type and dimensions
(4) One more minor: you are using `test_case_description` and a `Simd by`. may be just state that's a context explicitly?

So, as a possible suggestion:
```suggestion
    std::string log_msg("Failed for simd<");

    log_msg += m_data_type + "," + std::to_string(NumElems) + ">";
    log_msg += ", with context: " + ContextT::get_description();
    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
    log_msg += ", expected: " + std::to_string(m_expected_val);
    log_msg += ", at index: " + std::to_string(m_index);
```

Moreover, I have a feeling we don't need an index at all.. It has a different meaning for the `simd<T,1>` tests. Wouldn't it be a bit strange to retrieve something like this:
`Failed for simd<int,1> ... at index 4`. What do you think?

Reply:
Perfect idea! :)
This comment was applied [here](https://github.com/intel/llvm-test-suite/pull/614/commits/f5fcba9f39f6f63e6e4c8d2de3b4ea8cc64a0444#diff-a1a30a06ba7da50ce8e13a9af9caec10d385910429792a1de956a7f15cc3d646L57-L61)